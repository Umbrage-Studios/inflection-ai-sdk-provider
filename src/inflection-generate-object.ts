import { z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import {
  LanguageModelV1,
  LanguageModelV1Prompt,
  NoContentGeneratedError,
} from "@ai-sdk/provider";
import { InflectionChatLanguageModel } from "./inflection-chat-language-model";

export type GenerateObjectOptions = {
  model: LanguageModelV1;
  prompt: LanguageModelV1Prompt;
  output?: "object" | "array" | "enum" | "no-schema";
  schema?: z.ZodType;
  enum?: string[];
  maxTokens?: number;
  temperature?: number;
  topP?: number;
  stopSequences?: string[];
};

export async function generateObject({
  model,
  prompt,
  output = "object",
  schema,
  enum: enumValues,
  maxTokens,
  temperature,
  topP,
  stopSequences,
}: GenerateObjectOptions) {
  // Validate that the model is an Inflection model
  if (!(model instanceof InflectionChatLanguageModel)) {
    throw new Error("Model must be an Inflection AI model");
  }

  // Validate output strategy requirements
  if (output === "enum" && !enumValues?.length) {
    throw new Error(
      "Enum values must be provided when using enum output strategy"
    );
  }

  if ((output === "object" || output === "array") && !schema) {
    throw new Error(
      "Schema must be provided when using object or array output strategy"
    );
  }

  // Convert schema to JSON Schema if provided
  let jsonSchema: string | undefined;
  if (schema) {
    const converted = zodToJsonSchema(schema);
    jsonSchema = JSON.stringify(converted);
  }

  // Prepare the system message based on output strategy
  let systemMessage = "";
  switch (output) {
    case "object":
      systemMessage = schema
        ? `You must respond with a valid JSON object that strictly conforms to this schema: ${jsonSchema}. Do not include any explanatory text, only output valid JSON.`
        : "You must respond with a valid JSON object. Do not include any explanatory text, only output valid JSON.";
      break;
    case "array":
      systemMessage = schema
        ? `You must respond with a JSON array where each element strictly conforms to this schema: ${jsonSchema}. Do not include any explanatory text, only output valid JSON.`
        : "You must respond with a JSON array. Do not include any explanatory text, only output valid JSON.";
      break;
    case "enum":
      // We can safely use ! here because we validated enumValues exists above
      const validValues = enumValues!.join(", ");
      systemMessage = `You must respond with exactly one of these values: ${validValues}. Do not include any explanatory text or quotes, only output the exact value.`;
      break;
    case "no-schema":
      systemMessage =
        "You must respond with valid JSON. Do not include any explanatory text, only output valid JSON.";
      break;
  }

  // Add system message to prompt
  const fullPrompt: LanguageModelV1Prompt = [
    { role: "system", content: systemMessage },
    ...prompt,
  ];

  try {
    // Make the API call
    const response = await model.doGenerate({
      inputFormat: "prompt",
      mode: {
        type: "regular",
      },
      prompt: fullPrompt,
      maxTokens,
      temperature,
      topP,
      stopSequences,
      responseFormat: output === "enum" ? undefined : { type: "json" },
    });

    if (!response.text) {
      throw new NoContentGeneratedError({
        message: "No content was generated",
      });
    }

    // Parse and validate the response
    let parsedResponse: unknown;

    if (output === "enum") {
      // For enum output, use the raw text
      parsedResponse = response.text.trim();
    } else {
      // For other outputs, parse as JSON
      try {
        parsedResponse = JSON.parse(response.text);
      } catch (error) {
        throw new NoContentGeneratedError({
          message: `Failed to parse JSON: ${error instanceof Error ? error.message : "Unknown error"}`,
        });
      }
    }

    // Validate against schema if provided
    if (schema) {
      try {
        if (output === "array") {
          // For array output, wrap the schema in an array
          const arraySchema = z.array(schema);
          parsedResponse = arraySchema.parse(parsedResponse);
        } else {
          parsedResponse = schema.parse(parsedResponse);
        }
      } catch (error) {
        throw new NoContentGeneratedError({
          message: `Schema validation failed: ${error instanceof Error ? error.message : "Unknown error"}`,
        });
      }
    }

    // For enum output, validate against enum values
    if (output === "enum") {
      const enumSchema = z.enum(enumValues as [string, ...string[]]);
      try {
        parsedResponse = enumSchema.parse(parsedResponse);
      } catch (error) {
        throw new NoContentGeneratedError({
          message: `Invalid enum value: ${error instanceof Error ? error.message : "Unknown error"}`,
        });
      }
    }

    return {
      object: parsedResponse,
      response: response.response,
      usage: response.usage,
    };
  } catch (error) {
    if (error instanceof NoContentGeneratedError) {
      throw error;
    }
    throw new NoContentGeneratedError({
      message:
        error instanceof Error ? error.message : "Unknown error occurred",
    });
  }
}
