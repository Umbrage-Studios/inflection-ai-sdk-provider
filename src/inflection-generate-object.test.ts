import { expect, describe, it, beforeEach } from "vitest";
import { z } from "zod";
import { createTestServer } from "@ai-sdk/provider-utils/test";
import { createInflection } from "./inflection-provider";
import { generateObject } from "./inflection-generate-object";
import { NoContentGeneratedError } from "@ai-sdk/provider";

const INFERENCE_URL =
  "https://layercake.pubwestus3.inf7ks8.com/external/api/inference";

const server = createTestServer({
  [INFERENCE_URL]: {
    response: {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: '{"recipe":{"name":"Classic Lasagna","ingredients":[{"name":"lasagna noodles","amount":"1 package"},{"name":"ground beef","amount":"1 pound"}],"steps":["Boil noodles","Brown meat"]}}',
      },
    },
  },
});

const provider = createInflection({
  apiKey: "test-api-key",
  baseURL: INFERENCE_URL,
});

describe("generateObject", () => {
  beforeEach(() => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: '{"recipe":{"name":"Classic Lasagna","ingredients":[{"name":"lasagna noodles","amount":"1 package"},{"name":"ground beef","amount":"1 pound"}],"steps":["Boil noodles","Brown meat"]}}',
      },
    };
  });

  it("should generate an object using a schema", async () => {
    const recipeSchema = z.object({
      recipe: z.object({
        name: z.string(),
        ingredients: z.array(
          z.object({
            name: z.string(),
            amount: z.string(),
          })
        ),
        steps: z.array(z.string()),
      }),
    });

    const result = await generateObject({
      model: provider.chat("inflection_3_productivity"),
      schema: recipeSchema,
      prompt: [
        {
          role: "user",
          content: [{ type: "text", text: "Generate a lasagna recipe" }],
        },
      ],
    });

    expect(result.object).toMatchSnapshot();
  });

  it("should generate an array of objects", async () => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: '[{"name":"Warrior","class":"warrior","description":"A mighty fighter"},{"name":"Mage","class":"mage","description":"A powerful spellcaster"}]',
      },
    };

    const heroSchema = z.object({
      name: z.string(),
      class: z.string(),
      description: z.string(),
    });

    const result = await generateObject({
      model: provider.chat("inflection_3_productivity"),
      output: "array",
      schema: heroSchema,
      prompt: [
        {
          role: "user",
          content: [
            {
              type: "text",
              text: "Generate 2 hero descriptions for a fantasy game",
            },
          ],
        },
      ],
    });

    expect(result.object).toMatchSnapshot();
  });

  it("should handle enum output", async () => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: "sci-fi",
      },
    };

    const result = await generateObject({
      model: provider.chat("inflection_3_productivity"),
      output: "enum",
      enum: ["action", "comedy", "drama", "horror", "sci-fi"],
      prompt: [
        {
          role: "user",
          content: [
            {
              type: "text",
              text: 'Classify the genre of this movie plot: "A group of astronauts travel through a wormhole in search of a new habitable planet for humanity."',
            },
          ],
        },
      ],
    });

    expect(result.object).toBe("sci-fi");
  });

  it("should handle no-schema output", async () => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: '{"recipe":{"name":"Lasagna"}}',
      },
    };

    const result = await generateObject({
      model: provider.chat("inflection_3_productivity"),
      output: "no-schema",
      prompt: [
        {
          role: "user",
          content: [{ type: "text", text: "Generate a simple recipe" }],
        },
      ],
    });

    expect(result.object).toMatchSnapshot();
  });

  it("should throw NoContentGeneratedError for invalid JSON", async () => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: "This is not JSON",
      },
    };

    const recipeSchema = z.object({
      recipe: z.object({
        name: z.string(),
      }),
    });

    await expect(
      generateObject({
        model: provider.chat("inflection_3_productivity"),
        schema: recipeSchema,
        prompt: [
          {
            role: "user",
            content: [{ type: "text", text: "Generate a recipe" }],
          },
        ],
      })
    ).rejects.toThrow(NoContentGeneratedError);
  });

  it("should throw NoContentGeneratedError for schema validation failure", async () => {
    server.urls[INFERENCE_URL].response = {
      type: "json-value",
      body: {
        created: 1714688002.0557644,
        text: '{"wrong":"structure"}',
      },
    };

    const recipeSchema = z.object({
      recipe: z.object({
        name: z.string(),
      }),
    });

    await expect(
      generateObject({
        model: provider.chat("inflection_3_productivity"),
        schema: recipeSchema,
        prompt: [
          {
            role: "user",
            content: [{ type: "text", text: "Generate a recipe" }],
          },
        ],
      })
    ).rejects.toThrow(NoContentGeneratedError);
  });

  it("should throw error for enum output without enum values", async () => {
    await expect(
      generateObject({
        model: provider.chat("inflection_3_productivity"),
        output: "enum",
        prompt: [
          { role: "user", content: [{ type: "text", text: "Classify this" }] },
        ],
      })
    ).rejects.toThrow("Enum values must be provided");
  });

  it("should throw error for object output without schema", async () => {
    await expect(
      generateObject({
        model: provider.chat("inflection_3_productivity"),
        output: "object",
        prompt: [
          {
            role: "user",
            content: [{ type: "text", text: "Generate something" }],
          },
        ],
      })
    ).rejects.toThrow("Schema must be provided");
  });
});
