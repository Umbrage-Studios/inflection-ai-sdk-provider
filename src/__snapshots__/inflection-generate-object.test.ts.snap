// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateObject > should generate an array of objects 1`] = `
[
  {
    "class": "warrior",
    "description": "A mighty fighter",
    "name": "Warrior",
  },
  {
    "class": "mage",
    "description": "A powerful spellcaster",
    "name": "Mage",
  },
]
`;

exports[`generateObject > should generate an object using a schema 1`] = `
{
  "recipe": {
    "ingredients": [
      {
        "amount": "1 package",
        "name": "lasagna noodles",
      },
      {
        "amount": "1 pound",
        "name": "ground beef",
      },
    ],
    "name": "Classic Lasagna",
    "steps": [
      "Boil noodles",
      "Brown meat",
    ],
  },
}
`;

exports[`generateObject > should handle no-schema output 1`] = `
{
  "recipe": {
    "name": "Lasagna",
  },
}
`;
